### PR TITLE
Implement Ballerina Scan Tool for Static Code Analysis

### PR DESCRIPTION
**Pull request by Xenowa**
## Purpose
The purpose of this PR is to develop a Ballerina tool that performs static code analysis for Ballerina projects for identifying potential bugs, security vulnerabilities, and style violations and report analysis issues to static code analysis platforms like SonarQube.

Fixes [#42256](https://github.com/ballerina-platform/ballerina-lang/issues/42256#issue-2162985166)

## Goals
- Creating a Ballerina scan tool for performing core static code analysis based on language specific rules
- Extending the Ballerina scan tool for reporting analysis issues to static code analysis platforms
- Extending the Ballerina scan tool for introducing additional library specific rules and external rules

## Approach
The initial release introduces the `bal scan` command. Current implementation uses the project API to retrieve syntax trees of all Ballerina files in a project for performing core analysis and generating an analysis report. However the generated report will have no issues as core rules are not implemented yet. As core analysis rules are reviewed and implemented, the scan tool will progressively gain the ability to identify and report potential issues and further features will be added iteratively as outlined in task [#42260](https://github.com/ballerina-platform/ballerina-lang/issues/42260#issue-2163836547)

## Check List
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support
- [ ] Added necessary tests
    - [ ] Unit Tests
    - [ ]  Spec Conformance Tests
    - [ ] Integration Tests
    - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage
- [ ]  Added necessary documentation
    - [ ]  API documentation
    - [ ]  Module documentation in Module.md files
    - [ ]  Ballerina By Examples

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Ensure all Java files use LF.
+*.java eol=lf
```
Comment (SasinduDilshara) [inline .gitattributes:2]:
```suggestion
*.java eol=lf

```

File: .github/workflows/build-executor.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      - name: Grant execute permission for gradlew
+        run: chmod +x gradlew
+
+      - name: Build with Gradle
+        env:
+          packageUser: ${{ github.actor }}
+          packagePAT: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          ./gradlew build
```
Comment (SasinduDilshara) [inline .github/workflows/build-executor.yml:43]:
```suggestion
          ./gradlew build
          
```

File: .gitignore [hunk 3]
```diff
@@ -1,24 +1,64 @@
...
+/nbbuild/
+/dist/
+/nbdist/
+/.nb-gradle/
+
+### VS Code ###
+.vscode/
+
+### Mac OS ###
+.DS_Store
```
Comment (SasinduDilshara) [inline .gitignore:64]:
```suggestion
.DS_Store

```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,84 @@
...
+    ```
+
+3. Run analysis and save analysis results in specified directory.
+    ```bash
+    bal scan --target-dir="results"
+    ```
+4. Run analysis and generate an analysis report.
+    ```bash
+    bal scan --scan-report
+    ```
```
Comment (SasinduDilshara) [inline README.md:72]:
```suggestion
    ```
```

Comment (SasinduDilshara) [inline README.md:72]: We can remove this extra line

File: ScanCommand/src/main/java/io/ballerina/scan/Issue.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        this.ruleID = ruleID;
+        this.message = message;
+        this.issueType = issueType;
+        this.type = type;
+        this.fileName = fileName;
+        this.reportedFilePath = reportedFilePath;
+    }
+
+    public int getStartLine() {
+
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/Issue.java:59]:
```suggestion
```

File: ScanCommand/src/main/java/io/ballerina/scan/InternalReporter.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        if (issuesFilePath != null && INBUILT_RULES.containsKey(ruleID)) {
+            Issue issue = new Issue(startLine,
+                    startLineOffset,
+                    endLine,
+                    endLineOffset,
+                    ruleID,
+                    message,
+                    issueType,
+                    type,
+                    moduleName + "/" + documentName,
```
Comment (SasinduDilshara) [inline ScanCommand/src/main/java/io/ballerina/scan/InternalReporter.java:66]: File separator is depend on the OS. So it is better to not hardcode it.

File: ScanCommand/src/main/java/io/ballerina/scan/utilities/ScanUtils.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+                    target = new Target(targetDirectory);
+                } else {
+                    target = new Target(project.targetDir());
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        return target;
```
Comment (SasinduDilshara) [inline ScanCommand/src/main/java/io/ballerina/scan/utilities/ScanUtils.java:107]: You can remove the `target` variable and return in each scenario separately.

File: gradle.properties [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Scan tool properties
+scanToolVersion=1.0.0
```
Comment (SasinduDilshara) [inline gradle.properties:2]: Fix these `^M`

File: ScanCommand/src/main/java/io/ballerina/scan/ProjectAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                    analyzeDocument(project, module, documentId, internalScannerContext);
+                });
+            });
+        }
+
+        // return the detected issues
+        return allIssues;
+    }
+
+    public void analyzeDocument(Project currentProject,
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/ProjectAnalyzer.java:76]: Not sure whether this is the expected formatting for Java function definitions. Can you check

File: settings.gradle [hunk 10]
```diff
@@ -0,0 +1,2 @@
+rootProject.name = 'static-code-analysis-tool'
+include 'ScanCommand'
```
Comment (keizer619) [inline settings.gradle:2]:
```suggestion
include 'ScanCommand'

```

File: ScanCommand/tool-scan/tool_scan.bal [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+public function main() {
+    io:println("Ballerina Scan tool");
+}
```
Comment (keizer619) [inline ScanCommand/tool-scan/tool_scan.bal:20]:
```suggestion
}

```

**Comment by Xenowa** [comment]
Since this PR is too large, will be breaking it down to smaller parts and send them

**Pull request closed without merging**